### PR TITLE
API Gateway Policy Effect can only be `Allow` or `Deny` in @types/aws-lambda

### PR DESCRIPTION
**Pull request by ataylorme**
API Gateway Policy Effect can only be `Allow` or `Deny` as mentioned in [the API Gateway Control Access Policy Language documentation](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-control-access-policy-language-overview.html)

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

### CODE REVIEW

**Comment by G-Rath** [comment]
@ataylorme I'm happy with this change - just add some tests and we should be good to go

**Comment by ataylorme** [comment]
Yeah, I’ll clean up the tests when I have more time. Thanks for being open to the change! I ran into an issue with lowercase allow/deny erroring so I’m hoping this will save others in the future

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/aws-lambda/test/iot-authorizer-tests.ts
1. types/aws-lambda/trigger/api-gateway-authorizer.d.ts
1. types/aws-lambda/test/api-gateway-tests.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 4d6d28b02a92adcefa212ea90f4597a2932fe26d
</p>

**Comment by ataylorme** [comment]
@G-Rath the tests has a `createPolicyDocument` function that referenced `str`. It wasn't clear to me what the value of that would be, so I modified the function to accept an `effect` parameter that defaults to `Deny`.

I'm also open to ideas of how to test both `Allow` and `Deny` policy responses from the API Gateway and IoT Authorizer handlers

**Comment by ataylorme** [comment]
> In createPolicyDocument add tests with an otherwise valid statement assignment + different values for Effect including both Allow and Deny, "classic" low values like allow, and some completely arbitrary strings.

Added in [4241c18](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69715/commits/4241c18a1be44324ff20520380789a40d21cb607)

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/aws-lambda/test/iot-authorizer-tests.ts
1. types/aws-lambda/test/api-gateway-tests.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 4241c18a1be44324ff20520380789a40d21cb607
</p>

**Review by darbio — APPROVED** [review]

**Comment by darbio** [comment]
Love your work. Thank you!

**Comment by ataylorme** [comment]
> Love your work. Thank you!

Thank you and @G-Rath for the quick feedback and turn-around and all the maintainers. These types have helped me a lot so happy to contribute back ❤️.

**Review by G-Rath — APPROVED** [review]

**Comment by ataylorme** [comment]
Seems like @typescript-bot thinks this still needs maintainer review. Let me know if I need to do anything else.

**Review by darbio — COMMENTED** [review]
Reapproved

**Comment by darbio** [comment]
Because this is such a popular package, DT sometimes has an extra layer of approvals. They usually get to it in a few days.

**Comment by ataylorme** [comment]
Okay, thanks for the info

File: types/aws-lambda/trigger/api-gateway-authorizer.d.ts
Comment (simonbuchan) [inline types/aws-lambda/trigger/api-gateway-authorizer.d.ts:224]: Could you pull this out to a `export type StatementEffect =`?

**Review by simonbuchan — APPROVED** [review]
I'm a little concerned about plausibly breaking changes for valid code, but I think this one is an easy and rare enough fix to be well worth it.

**Comment by darbio** [comment]
> I'm a little concerned about plausibly breaking changes for valid code, but I think this one is an easy and rare enough fix to be well worth it.

I agree, however I'm not against narrowing the types to make them more accurately reflect the implementation. 

Narrowing the type here may (will) break someone's code (for instance if they are deriving from a function that returns string), however tsc will give a helpful error at compile time which makes it a trivial fix. 

Especially given the type string is _technically_ not valid and this is a change for the better.

File: types/aws-lambda/trigger/api-gateway-authorizer.d.ts
Comment (ataylorme) [reply types/aws-lambda/trigger/api-gateway-authorizer.d.ts]: Added in [1825bbe](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69715/commits/1825bbefd3023f44a41b083f5716ab14dbf4f3ee)

**Review by G-Rath — APPROVED** [review]

**Review by simonbuchan — APPROVED** [review]

**Review by darbio — APPROVED** [review]
Approved. Thank you.

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by G-Rath** [comment]
Ready to merge

**Comment by ataylorme** [comment]
Ready to merge

**Pull request merged**
